### PR TITLE
M3X-720: Enhance ins/del wrapping on atomic tags while diffing

### PR DESCRIPTION
**Pull request by ragoolkr**
[Ticket](https://matrixreq.atlassian.net/browse/M3X-720) 

This pull request introduces significant improvements to the HTML diffing logic, focusing on more robust handling of atomic tags.

### **ISSUE**
image

In html-diff when we have the tag with data-html-diff-id we consider them atomic and wrap them with ins/del. we decide opening and close of the atomic tag based on the tag name.

**Sometimes we might have a structure where** 
```html
<span data-htmldiff-id="testing">
  <span class="truncate">one</span>
  <span class="truncate">two</span>
</span>
``` 
then diffing algorithm finds the `data-htmldiff-id `starts the wrapping process it looks for the atomic tag which is `<span> `then it starts looking for the closing tag `</span>` but in our structure since it also has a child `<span>` the algorithm considers child's closing `</span>` as the parent one and produces something like this 

```html
<ins>
 <span data-htmldiff-id="testing">
   <span class="truncate">one</span>
</ins>
   <span class="truncate">two</span>
</span> // orphan tag
``` 
### **FIX**
So now we track how many tags with same name currently open inside the atomic token old code closed atomic tag at first </span> nested spans cut early broken markup.we count open/close depth, closes only when depth at 0 → whole nested block stays one balanced token.

working version

**Core logic improvements:**

* Improved handling of atomic tags (e.g., `<iframe>`, `<script>`, and custom tags with `data-htmldiff-id`), ensuring each atomic tag is wrapped independently in diffs, and supporting nested and adjacent atomic tags.

Fix counts open/close depth, closes only at depth 0 → whole nested block stays one balanced token.

**Testing and CI:**

* Added and updated tests for edge cases, atomic tag handling and HTML tokenization to ensure correctness and prevent regressions.

**Package metadata:**

* Updated `package.json` to reflect new ownership (`@matrixreq/htmldiff`), version bump to 1.4.0, and repository links.

These changes collectively make the diffing library more reliable for complex HTML structures and easier to maintain and test.

### CODE REVIEW

File: js/htmldiff.js
Comment (sirenevenkii) [inline js/htmldiff.js:1]: the naming is really weird and make things hard to understand. 

`lastClosedTag` - if it's (get?)lastClosedTag, why we check whether it's opening or closing tag later? should we just have `isClosingTagOf` and `isOpeningTagOf` (not sure about that, but I would like to simplify and make it clearer a bit)

File: js/htmldiff.js [hunk 1]
```diff
@@ -202,14 +218,23 @@
...
+                    if (isEndOfTag(char)){
+                        var closedTag = lastClosedTag(currentWord);
+                        if (closedTag && isClosingTagOf(closedTag, currentAtomicTag)){
+                            currentAtomicTagDepth--;
+                            if (currentAtomicTagDepth <= 0){
+                                words.push(createToken(currentWord));
+                                currentWord = '';
+                                currentAtomicTag = '';
+                                currentAtomicTagDepth = 0;
+                                mode = 'char';
```
Comment (sirenevenkii) [inline js/htmldiff.js:232]: should it be similar to this (from just above)? add a test if it should 

```
else if (isEndOfTag(char)){
                        currentWord += '>';
                        words.push(createToken(currentWord));
                        currentWord = '';
                        if (isWhitespace(char)){
                            mode = 'whitespace';
                        } else {
                            mode = 'char';
                        }
```

Comment (ragoolkr) [reply js/htmldiff.js:232]: not needed since the block is already inside `if (isEndOfTag(char))` which confirms `char === '>'`
so this will never be true inside here` if (isWhitespace(char))`.

File: js/htmldiff.js
Comment (ragoolkr) [reply js/htmldiff.js:1]: yes `lastClosedTag` can be removed we can directly check for the stray `> a` inside the `isOpeningTagOf` and  `isClosingTagOf` and have only these two.

File: js/htmldiff.js [hunk 2]
```diff
@@ -88,18 +88,26 @@
...
-     * @param {string} tag The ending tag to look for.
-     *
-     * @return {boolean} True if the word is now a complete token (including the end tag),
-     *    false otherwise.
+     * @return {boolean} True if word ends with an opening (non-self-closing) tag for the given
+     *    atomic tag name.
+     */
+    function isOpeningTagOf(word, tag){
+        var tagText = word.substring(word.lastIndexOf('<'));
+        if (tagText.indexOf('>') !== tagText.length - 1) return false;
```
Comment (sirenevenkii) [inline js/htmldiff.js]: do not use inline conditions. same for closing tag

File: js/htmldiff.js [hunk 3]
```diff
@@ -88,18 +88,26 @@
...
-     * Checks if the current word is the end of an atomic tag (i.e. it has all the characters,
-     * except for the end bracket of the closing tag, such as '<iframe></iframe').
-     *
-     * @param {string} word The characters of the current token read so far.
-     * @param {string} tag The ending tag to look for.
-     *
-     * @return {boolean} True if the word is now a complete token (including the end tag),
-     *    false otherwise.
+     * @return {boolean} True if word ends with an opening (non-self-closing) tag for the given
+     *    atomic tag name.
```
Comment (sirenevenkii) [inline js/htmldiff.js]: it's not about specifically atomic tags, isn't? same for `isClosingTagOf` comment

File: js/htmldiff.js [hunk 4]
```diff
@@ -88,18 +88,26 @@
         return result && result[1];
     }
 
+    // inspect the last tag in word (from its opening '<'); to an earlier '>' in that
+    // slice means loose text e.g. "a > b" in a <script>, not a tag, so they bail out.
```
Comment (sirenevenkii) [inline js/htmldiff.js]: why those comment are inline?

Comment (sirenevenkii) [reply js/htmldiff.js]: paraphrase it, I'm struggling to understand what you mean here

**Review by sirenevenkii — APPROVED** [review]
looks good. great improvement. 
please, clean up the comments and styling before merging.
